subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveBuildCacheKeyExecuter.java

```diff
@@ -144,6 +150,145 @@ public String apply(HashCode input) {
+            public String getPath() {
+                return path;
+            }
+
+            @Override
+            public String getHash() {
+                return hash.toString();
+            }
+
+            @Override
+            public boolean preVisitDirectory(String path, String name) {
+                this.path = path;
+                this.name = name;
+                this.hash = null;
+
+                visitor.preDirectory(this);
+
+                return true;
+            }
+
+            @Override
+            public void visit(String path, String name, FileContentSnapshot content) {
+                FileContentSnapshot snapshot = snapshot(path);
+                if (snapshot == null) {
+                    return;
+                }
+
+                this.path = path;
+                this.name = name;
+                this.hash = snapshot.getContentMd5();
+
+                visitor.file(this);
+            }
+
+            private FileContentSnapshot snapshot(String path) {
+                NormalizedFileSnapshot normalizedFileSnapshot = normalizedSnapshots.get(path);
+                if (normalizedFileSnapshot == null) {
+                    // This file was ignored by normalization
+                    return null;
+                }
+                FileContentSnapshot snapshot = normalizedFileSnapshot.getSnapshot();
+                if (snapshot == null) {
+                    throw new IllegalStateException("snapshot is null : " + path);
+                }
+                return snapshot;
+            }
+
+            @Override
+            public void postVisitDirectory() {
+                visitor.postDirectory();
+            }
+        }
+
+
+        @Override
+        public void visitInputFileProperties(InputFilePropertyVisitor visitor) {
+            State state = new State(visitor);
+            ImmutableSortedMap<String, FileCollectionSnapshot> inputFiles = key.getInputs().getInputFiles();
+            if (inputFiles == null) {
+                return;
+            }
+            for (Map.Entry<String, FileCollectionSnapshot> entry : inputFiles.entrySet()) {
+                FileCollectionSnapshot fileCollectionSnapshot = entry.getValue();
+
+                state.propertyName = entry.getKey();
+                state.propertyHash = fileCollectionSnapshot.getHash();
+                state.propertyNormalizationStrategyName = fileCollectionSnapshot.getNormalizationStrategyName();
+                state.normalizedSnapshots = fileCollectionSnapshot.getSnapshots();
+
+                visitor.preProperty(state);
+                for (PhysicalSnapshot tree : fileCollectionSnapshot.getRoots()) {
+                    if (tree.getType() == FileType.Missing) {
+                        continue;
+                    }
+
+                    state.path = tree.getAbsolutePath();
+                    state.name = tree.getName();
+                    state.hash = null;
+
+                    visitor.preRoot(state);
```

Comment:
On `master`, there is now only an `accept` method on `FileCollectionSnapshot`. That method would visit all roots in order. The visitor here would need to make sure it detects when we switch roots. Do you think that is feasible?

For checking if something is a root the visitor would need to keep track of the depth it currently is by tracking `preVisitDirectory`/`postVisitDirectory` calls.